src/machine/machine_rp2040-rtc.go

```diff
@@ -0,0 +1,256 @@
+	irqSetup0 volatile.Register32
+	irqSetup1 volatile.Register32
+	rtc1      volatile.Register32
+	rtc0      volatile.Register32
+	intR      volatile.Register32
+	intE      volatile.Register32
+	intF      volatile.Register32
+	intS      volatile.Register32
+}
+
+var RTC = (*rtcType)(unsafe.Pointer(rp.RTC))
+
+var rtcAlarmRepeats bool
+var rtcCallback func()
+
+var ErrRtcNotRunning = errors.New("RTC not running")
+var ErrRtcInvalidTime = errors.New("invalid time for RTC")
+
+func (rtc *rtcType) running() bool {
+	return rtc.ctrl.HasBits(rp.RTC_CTRL_RTC_ACTIVE)
+}
+
+func (rtc *rtcType) init() {
+	// Get clk_rtc freq and make sure it is running
+	rtcFreq := configuredFreq[clkRTC]
+	if rtcFreq == 0 {
+		panic("rtc freq is zero")
+	}
+
+	// Take rtc out of reset now that we know clk_rtc is running
+	resetBlock(rp.RESETS_RESET_RTC)
+	unresetBlockWait(rp.RESETS_RESET_RTC)
+
+	// Set up the 1 second divider.
+	// If rtc_freq is 400 then clkdiv_m1 should be 399
+	rtcFreq -= 1
+
+	// Check the freq is not too big to divide
+	if rtcFreq > rp.RTC_CLKDIV_M1_CLKDIV_M1_Msk {
+		panic("rtc freq is too big to divide")
+	}
+
+	// Write divide value
+	rtc.clkDivM1.Set(rtcFreq)
+}
+
+func (rtc *rtcType) SetTime(t RtcTime) error {
+	if !t.IsValid() {
+		return ErrRtcInvalidTime
+	}
+
+	// Disable RTC and wait while it is still active
+	rtc.ctrl.Set(0)
+	for rtc.running() {
+	}
+
+	rtc.setup0.SetBits(uint32(t.Year) << rp.RTC_SETUP_0_YEAR_Pos)
+	rtc.setup0.SetBits(uint32(t.Month) << rp.RTC_SETUP_0_MONTH_Pos)
+	rtc.setup0.SetBits(uint32(t.Day) << rp.RTC_SETUP_0_DAY_Pos)
+
+	rtc.setup1.SetBits(uint32(t.Dotw) << rp.RTC_SETUP_1_DOTW_Pos)
+	rtc.setup1.SetBits(uint32(t.Hour) << rp.RTC_SETUP_1_HOUR_Pos)
+	rtc.setup1.SetBits(uint32(t.Min) << rp.RTC_SETUP_1_MIN_Pos)
+	rtc.setup1.SetBits(uint32(t.Sec) << rp.RTC_SETUP_1_SEC_Pos)
+
+	// Load setup values into rtc clock domain
+	rtc.ctrl.SetBits(rp.RTC_CTRL_LOAD)
+
+	// Enable RTC and wait for it to be running
+	rtc.ctrl.SetBits(rp.RTC_CTRL_RTC_ENABLE)
+	for !rtc.running() {
+	}
+
+	return nil
+}
+
+func (rtc *rtcType) GetTime() (t RtcTime, err error) {
+	// Make sure RTC is running
+	if !rtc.running() {
+		return RtcTime{}, ErrRtcNotRunning
+	}
+
+	// Note: RTC_0 should be read before RTC_1
+	rtc_0 := rtc.rtc0.Get()
+	rtc_1 := rtc.rtc1.Get()
+
+	t = RtcTime{
+		Dotw:  int8((rtc_0 & rp.RTC_RTC_0_DOTW_Msk) >> rp.RTC_RTC_0_DOTW_Pos),
+		Hour:  int8((rtc_0 & rp.RTC_RTC_0_HOUR_Msk) >> rp.RTC_RTC_0_HOUR_Pos),
+		Min:   int8((rtc_0 & rp.RTC_RTC_0_MIN_Msk) >> rp.RTC_RTC_0_MIN_Pos),
+		Sec:   int8((rtc_0 & rp.RTC_RTC_0_SEC_Msk) >> rp.RTC_RTC_0_SEC_Pos),
+		Year:  int16((rtc_1 & rp.RTC_RTC_1_YEAR_Msk) >> rp.RTC_RTC_1_YEAR_Pos),
+		Month: int8((rtc_1 & rp.RTC_RTC_1_MONTH_Msk) >> rp.RTC_RTC_1_MONTH_Pos),
+		Day:   int8((rtc_1 & rp.RTC_RTC_1_DAY_Msk) >> rp.RTC_RTC_1_DAY_Pos),
+	}
+
+	return t, nil
+}
+
+// void rtc_set_alarm(datetime_t *t, rtc_callback_t user_callback) {
+func (rtc *rtcType) SetAlarm(t RtcTime, callback func()) {
```

Comment:
Should be documented that the callback will be in the context of an interrupt handler - ie no blocking, no memory allocation, etc?